### DISCUSSION TITLE
Conditional constraint

### DISCUSSION THREAD

**Discussion by gioperin** [do-mpc/do-mpc#193] [Q&A] [answered]
Hello, 

I am looking for help on how to code a conditional constraint. I put a minimal understandable example. I have two state variables, namely, the intensity and the deadline of a computing process:

`I_k = model.set_variable(var_type='_x', var_name='I_k')`
` D_k = model.set_variable(var_type='_x', var_name='D_k')`

My control output is the amount of intensity that I can process at time t:

`wrkld = model.set_variable(var_type='_u', var_name='wrkld')`

The two state equations evolve in this way:

`I_k_next = I_k - wrkld`
`D_k_next = D_k - tau`

where tau is a constant. That is, the deadline is like a timer. What I need is that if the deadline expires during the current slot, the `wrkld` is equal to the `I_k`, i.e., if we have no more time, we must process the residual computations. Currently, I am trying to reach this behavior with the Casadi if_else method. The problem is that it requires an else:

`forced = casadi.if_else(model.x['D_k'] <= tau, model.x['I_k'], 0)`
`mpc.set_nl_cons('forced_execution_1', forced, 0, soft_constraint=True, penalty_term_cons=delta_cost)`
`mpc.set_nl_cons('forced_execution_2', -forced, 0, soft_constraint=True, penalty_term_cons=delta_cost).`

Of course it doesn't work, because it (soft) constrains the control value to be equal to the residual intensity when the deadline is about to expire (desired), but also to be equal to 0 when this is not true (strongly undesired!).

How can I reach the desired behavior?

Thanks in advance for any help. If you need further information, just ask me!

Giovanni

**Comment by 4flixt** [comment, marked as answer]
Hi @gioperin,

I have been thinking a bit about your issue and I think the main problem is the switching condition of ``if_else`` which causes IPOPT some troubles.

To fix this issue I suggest in your case to use[ time-varying parameters](https://www.do-mpc.com/en/latest/FAQ.html#time-varying-parameters) to model the deadline since the dynamic of that "state" is trivial. 

Other than that I am not entirely sure why your are using soft-constraints for your problem? The constraint is definitely satisfied for ``model.x['D_k'] > tau`` 

what is your intended behaviour for ``model.x['D_k'] <= tau``?

Best
4flixt

P.S.: I have converted this issue to a discussion.

**Reply by gioperin** [reply]
Hi @4flixt

I will consider your suggestion of making `D_k` a time-varying parameter since it has no relations with the control variable. What I wanted to reach is that if `model.x['D_k'] <= tau`, then `model.u['wrkld'] = model.x['I_k']`. If it is not verified, no constraint should be set. Nonetheless, like it was confirmed by the Casadi support, the casadi.if_else() method requires an else. Thus, the workaround that I found was to play with the physical limits of my variable `model.u['wrkld']`. It may be of interest for someone else. Now I have this:

`mpc.bounds['lower', '_u', 'wrkld'] = 0`
`mpc.set_nl_cons('max_power', model.u['wrkld'], max_power, soft_constraint=True, penalty_term_cons=delta_cost)`
`mpc.set_nl_cons('wrkld_conservation', model.u['wrkld'] - model.x['I_k'], 0)`
`forced = casadi.if_else(model.x['D_k'] <= tau, model.x['I_k'], 0)`
`mpc.set_nl_cons('forced_execution', forced - model.u['wrkld'], 0)`

Basically, when I did not want any constraint on my control, I just repeated the constraint that the variable must be greater than 0, which is its lower limit. I would not know how to handle unbounded controls, though. I removed the soft constraint from forced execution, and left it only on the maximum power, so that the controller does not break.

A (not so) related question about `set_nl_cons()`, for which I could not find answers on the docs: the additional penalty goes inside the cost function with a linear or quadratic term? I think it is worth it to mention this in the docs.

Thank you so much for your time,
I will let all know if modeling `D_k` as a parameter stabilizes things.

Giovanni

**Reply by gioperin** [reply]
Dear @4flixt,

your suggestion was correct: using time-varying parameters for trivial states stabilizes a lot the output result (other than the modifications that I already did and explained in the previous comment).

Thanks a lot!

Giovanni
